### PR TITLE
Correct the frame timing in 'Rasterizer::Draw' when pipeline is more available

### PR DESCRIPTION
**Pull request by ColdPaleLight**
The current implementation always uses `resubmit_recorder` as the `frame_timings_recorder` for the next `Rasterizer::Draw` when `consume_result` is `PipelineConsumeResult::MoreAvailable`. However, in fact, we should use `resubmit_recorder` only when `should_resubmit_frame` is `true`, and for other cases, we should use the `frame_timing_recorder` corresponding to the next item in the pipeline.

In this PR, I introduced a new struct `LayerTreeItem` to store `layer_tree` and `frame_timings_recorder`, and push the `LayerTreeItem` into `Pipeline`. This ensures that the correct `frame_timings_recorder` is used in `Rasterizer::Draw`;

fix https://github.com/flutter/flutter/issues/96699

c.f.
https://github.com/flutter/engine/blob/98962eebf50f519340985371568db2f17b4ab49b/shell/common/rasterizer.cc#L226

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: shell/common/rasterizer.cc
Comment (ColdPaleLight) [inline shell/common/rasterizer.cc:163]: Since `frame_timings_recorder` is in the pipeline, it can't be accessed here. So I changed the code here back to before this PR https://github.com/flutter/engine/pull/26205 and added a trace in `DoDraw`. It looks like this will have an impact on the dev tools. Please let me know if you guys have any thoughts.

**Comment by zanderso** [comment]
From PR review triage: @chinmaygarde @iskakaushik is this ready for a review?

**Comment by chinmaygarde** [comment]
I don't have the cycles to review this ATM. I'll bring this up with Kaushik.

**Comment by ColdPaleLight** [comment]
friendly ping @iskakaushik

File: shell/common/rasterizer.cc
Comment (iskakaushik) [reply shell/common/rasterizer.cc:165]: This might impact the devtools, as it looks for specific frame number events corresponding to `::Draw`. @kenzieschmoll  do we still rely on the names of trace events or do we only care about the frame timings API?

**Comment by iskakaushik** [comment]
This change looks good to me overall, I will wait for Kenzie's response on this reg. devtools.

File: shell/common/rasterizer.cc
Comment (kenzieschmoll) [reply shell/common/rasterizer.cc:165]: We look for the 'frame_number' arg in the 'GPURasterizer::Draw' trace event and the 'Animator::BeginFrame' trace event to link timeline events to their matching frame number from the FrameTimings API. If this change removes the 'frame_number' argument, this will break DevTools.

Comment (iskakaushik) [reply shell/common/rasterizer.cc:165]: I filed https://github.com/flutter/flutter/issues/105140, once that is addressed, we can land this PR.

**Comment by zanderso** [comment]
I see the devtools change has landed. Is this one ready to go?

**Review by iskakaushik — APPROVED** [review]
LGTM

**Comment by iskakaushik** [comment]
@ColdPaleLight  thanks for the contribution.

**Pull request merged**
